subsystem/src/manager.rs

```diff
@@ -308,60 +282,84 @@ impl Manager {
         log::info!("Manager {} starting subsystems", self.name);
 
         // Run all the subsystem tasks.
-        let mut task_handles: Vec<_> = self.subsystem_tasks.into_iter().map(task::spawn).collect();
+        let subsystems: Vec<_> = self
+            .subsystems
+            .into_iter()
+            .map(|s| (s.name, task::spawn(s.task), s.shutdown_tx))
+            .collect();
 
         // Signal the manager is shut down so it does not wait for itself
         drop(self.shutting_down_tx);
 
-        // Wait for a subsystem to shut down
-        loop {
-            // We have to handle the empty case explicitly to avoid panic
-            let tasks_join_future = if task_handles.is_empty() {
-                std::future::pending().left_future()
-            } else {
-                select_all(task_handles).right_future()
-            };
-            // Wait for either the shutdown signal or task crash
-            tokio::select! {
-                (result, _, rest) = tasks_join_future => {
-                    task_handles = rest;
-                    match result {
-                        // Task terminated gracefully
-                        Ok(()) => continue,
-                        // Task terminated in an unexpected way
-                        Err(e) => {
-                            let msg =
-                                format!("Manager {}: error from a subsystem: {}", self.name, e);
-                            log::error!("{}", msg);
-                            panic!("{}", msg);
-                        }
-                    }
-                }
-                shut = self.shutting_down_rx.recv() => {
-                    if shut.is_none() {
-                        log::info!("Manager {}: all subsystems already down", self.name);
-                    }
-                }
-            }
-            break;
+        // Wait for the shutdown trigger.
+        if self.shutting_down_rx.recv().await.is_none() {
+            log::warn!("Manager {}: all subsystems already down", self.name);
         }
-
         log::info!("Manager {} shutting down", self.name);
+        // Drop the receiver in order to prevent blocking of subsystems.
+        drop(self.shutting_down_rx);
 
-        // Order all the remaining subsystems to shut down.
-        let _ = self.shutdown_request_tx.send(());
+        // Shut down the subsystems in the reverse order of creation.
+        for (name, handle, shutdown_tx) in subsystems.into_iter().rev() {
+            if let Err(()) = shutdown_tx.send(()) {
+                log::warn!("Manager {}: {name} subsystem is already down", self.name);
```

Comment:
Should this `continue` rather than fall through as this subsystem has already shutdown?

Reply:
I thought about that, but joining the handle allows to print a potential error message.